### DISCUSSION TITLE
How to solve: "baseline and build for xyz.class have different contents" ?

### DISCUSSION THREAD

**Discussion by fedejeanne** [eclipse-platform/eclipse.platform.releng.aggregator#2406] [Q&A] [answered]
Hello everyone,
I ran into these failed checks in JDT (under "_Compiler and API tools_") and I can't really understand what they mean and how to solve them:

```
baseline and build for org/eclipse/jdt/internal/ui/text/correction/QuickAssistProcessorUtil.class have different contents
```
and
```
baseline and build for org/eclipse/jdt/internal/corext/fix/StringConcatToTextBlockFixCore$StringBufferFinder.class have different contents
```

Here's the build: https://github.com/eclipse-jdt/eclipse.jdt.ui/pull/1671/checks?check_run_id=30751318486

I'm pretty sure I read something about it in the official documentation but it was some time ago and I can't find the link anymore.

I'm tempted to say that the solution is bumping the version of the plugin in its MANIFEST, in this case it would be: `org.eclipse.jdt.ui`, but I'm a bit thrown off because increasing it already shows an error locally:

![image](https://github.com/user-attachments/assets/c4948ef2-00c0-44cc-b5dc-9192ab045a14)

Any hints?

**Comment by merks** [comment, marked as answer]
The API baseline replacement stuff is quite fussy.   Firstly, it does not like it when a new artifact is produced that has the same artifact ID (fully qualified version) as in the baseline(s).  It also doesn't like a new artifact that has only a qualifier increment, so typically the first change to a bundle for a release cycle requires at least a micro/service increment (of 100).   But after that, it's still possible that something changes, e.g., the compiler produces new different bytes.  At this point there will again be complaints about the artifact being different but having the same version.  So I think there are conceptually kind two baselines, the API baseline from the most-recent release, and the baseline from the already-published/publicized I-Builds)   At this point just new qualifier is needed, not another micro/service increment.  That's generally done by editing the forceQualifierUpdate.txt with some description/link/whatever for why the qualifier needs to be different.  

So, in general, when something changes, the artifact needs a new version.  That version must be at least +100 from the API baseline and the qualifier needs to be bigger than the qualifier from any previous I-builds before the change...

Someone should correct me on anything I've said that isn't entirely accurate!

**Reply by laeubi** [reply]
> So I think there are conceptually kind two baselines, the API baseline from the most-recent release, and the baseline from the already-published/publicized I-Builds)

Yes we have two processes `baseline-compare` and `baseline-replace` (the build-baseline) **and** we have api-baseline (aka API tools).

The API baseline is always about **API changes** compared against the last release(s).
The build-baseline is about **content changes** compared to the last I-build.

The above error is from API tools and complains that the version micro version is increased (== not equal to 0) even if the the minor (e.g. 33) version was already incremented compared to the last release, or someone else already changed it from 0 --> 100 already.

Such error can also show up if you are not "up-to-date" with master, so often pull+rebase solve the problem already.

If not as mentioned, sometimes there needs to be a "touch", that means, the version stay the same but some file (usually a forceQualifierUpdate.txt) needs to be changed so we get a new qualifier change.

**Reply by fedejeanne** [reply]
Thank you both for the clarification. 
What I gather from your responses is:

1. The error `baseline and build for xyz.class have different contents` refers to the **build-baseline**
2. The error `The service version is increased unnecessarily...` (the one in the screenshot) refers to the **API baseline**
3. I "fixed" the **build-baseline** because I _touched_ a file but it happened to be the wrong file: the MANIFEST (and the POM).
4. Next time, I should _touch_ another file: _**forceQualifierUpdate**.txt_

> Such error can also show up if you are not "up-to-date" with master, so often pull+rebase solve the problem already.

I tried that but it didn't fix the issue mentioned in _1._

**Reply by iloveeclipse** [reply]
> Next time, I should touch another file: forceQualifierUpdate.txt

No, you shouldn't. This should only happen in very seldom cases where the code change from one bundle affects compilation **artifacts** of another bundle (without breaking source code compile) and other bundle has no source changes.

**Reply by merks** [reply]
Indeed, I don't ever recall needing to do this for a bundle.  For a feature, when most still included 3rd party bundles, an update to the target platform often required such forced qualifiers for the including features.  But we've mostly eliminated that...

**Reply by fedejeanne** [reply]
So if a _pull + rebase_ didn't solve the problem (I tried it) and _touching_ `forceUpdateQualifier.txt` isn't the right solution, what other options do I have if this problem happens again? I assume it's not safe to simply ignore the error (_i.e._ "leave it to the committer").

**Reply by iloveeclipse** [reply]
You should ask, it is hard to say which problem you had before and which might happen in the future. We permanently have infrastructure issues, tycho updates, github action changes, PDE bugs etc

**Reply by fedejeanne** [reply]
Good point, the infrastructure is also a changing environment so asking is never a bad idea.

Thank you all for the insights! I consider my question answered and I've learned something new in this discussion :-)
